### ISSUE TITLE
Try blocks give false errors (`type-mismatch` false positive)

### ISSUE THREAD

**Issue by Kixiron** [rust-lang/rust-analyzer#11843]
Rust Analyzer sometimes gives errors on try blocks where there shouldn't be any, e.g.
```rust
let _: Option<()> = try {
    let first_rhs = self.values.get(graph.input_source(sub.rhs()))?;

    let lhs_sub = *graph.cast_input_source::<Sub>(sub.lhs())?;
    let second_rhs = self.values.get(graph.input_source(lhs_sub.rhs()))?;

    let sum = first_rhs + second_rhs;
    let new_rhs = graph.constant(sum);
    self.values.add(new_rhs.value(), sum);

    graph.remove_inputs(sub.node());

    let value = graph.input_source(lhs_sub.lhs());
    graph.add_value_edge(value, sub.lhs());
    graph.add_value_edge(new_rhs.value(), sub.rhs());

    self.changes.inc::<"sub-sub">();
};
```
This code is correct (and rustc doesn't complain), but RA emits this error:
![image](https://user-images.githubusercontent.com/25047011/160657470-20f77708-3576-4b0a-97ac-436a3bae136a.png)

Changing it as suggested (adding a `Some(())` or `None` to the end of the block) is incorrect and (correctly) causes both rustc and RA to emit errors

**rust-analyzer version**: `bc08b8eff 2022-03-28 stable`

**rustc version**: (eg. output of `rustc -V`)
```
rustc 1.60.0-nightly (777bb86bc 2022-01-20)
binary: rustc
commit-hash: 777bb86bcdbc568be7cff6eeeaaf81a89b4aa50b
commit-date: 2022-01-20
host: x86_64-pc-windows-msvc
release: 1.60.0-nightly
LLVM version: 13.0.0
```

**relevant settings**: N/A

**Comment by cjhopman** [comment]
While it's not implemented, I'm pretty sure it's a recent regression that it's now showing errors here. We've used try_blocks for quite a while and haven't had this issue.

**Comment by cjhopman** [comment]
Yeah, I see this on 0.2.993, but not on 0.2.985.

**Comment by bjorn3** [comment]
We recently started showing type mismatch errors. Before that those weren't shown at all.

**Comment by peku33** [comment]
expiriencing the same issue

**Comment by lnicola** [comment]
They are disabled by default. Check if you're enabled experimental diagnostics or if you're on an older version.

**Comment by ndmitchell** [comment]
@lnicola I had pinned my Rust Analyzer extension due to a previous bug - removing the pin fixed the errors. Thanks!

**Comment by KapJI** [comment]
I have these errors on Windows even with experimental diagnostics disabled :(

RA version: ad6810e90 2022-06-06 stable
VS Code extension: 0.3.1083

**Comment by flodiebold** [comment]
@KapJI can you show an example?

**Comment by flodiebold** [comment]
Hmm yeah, that's happening because we have a quick fix in this case (the "wrap in Ok" suggestion). You could disable the diagnostic completely instead.

**Comment by ImUrX** [comment]
is this possible to PR? or does it need to reach stable?

**Comment by Veykril** [comment]
An implementation for this doesn't need to wait for try blocks to become stable

**Comment by HKalbasi** [comment]
What do you think about adding a `confidence level` to `InferenceResult`, decreasing it in presence of unresolved names, unknown types, unimplemented features, ... and suppressing `type-mismatch` for whole body if the confidence level is low? It seems most of the false positives of `type-mismatch` would be suppressed with this, and it may even help adding more diagnostics without fear of false positives, like `not-implemented` which reverted some times ago.

**Comment by flodiebold** [comment]
I would rather just implement support for try blocks.

In other cases, e.g. unresolved names and unknown types, the error type is supposed to act as an error suppressor. So we e.g. shouldn't actually emit type mismatches involving an error type, because we know there already was an error before.

**Comment by HKalbasi** [comment]
Supporting try blocks is a different problem. Try blocks won't be the last feature which r-a supports with a delay.

In other cases, it is basically an extreme solution for #6975. I'm not sure about unknown types, since they might be *caused* by a type mismatch, but I think suppressing in presence of unresolved names or incomplete expressions does more good than harm. If r-a knowledge is full, user can first fix the unresolved name and then fix the type mismatch errors, and if r-a knowledge is partial, then the type mismatch error has a good chance to be wrong as well.

**Comment by ndmitchell** [comment]
I write in a code base with a lot of features enabled. Of those, only try blocks cause RA to have fairly significant errors in inference, so I do think try blocks are slightly different from the other features that aren't yet supported.

**Comment by flodiebold** [comment]
> Supporting try blocks is a different problem. Try blocks won't be the last feature which r-a supports with a delay.

This approach would not be what we want long-term: Long-term I would expect that we can either support most features fast enough since we're not still catching up anymore, or at least detect new features and suppress errors for them specifically. For example, we could just infer the error type for try blocks and rely on follow-up error suppression. That would require the same effort as reducing the confidence level in your approach.

It's also not what we want short-term, because we want to see bugs and not paper over them. The type-mismatch diagnostic is already experimental anyway (except it shows up here because we heuristically show it when we have a quickfix, but that's something we could address in the heuristic), so I don't see the need to hide it in even more cases.
